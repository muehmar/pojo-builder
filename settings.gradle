rootProject.name = "pojo-builder"

include 'core'
include 'core-annotations'
include 'example'
include 'example-java17'

dependencyResolutionManagement {
    versionCatalogs {
        libs {
            version('pojobuilder', '1.6.0')
            library('pojobuilder-core', 'io.github.muehmar','pojo-builder').versionRef('pojobuilder')
            library('pojobuilder-annotations', 'io.github.muehmar','pojo-builder-annotations').versionRef('pojobuilder')

            library('codeGenerator', 'io.github.muehmar:code-generator:0.22.0')

            version('junit', '5.10.0')
            library('junit-api', 'org.junit.jupiter', 'junit-jupiter-api').versionRef('junit')
            library('junit-engine', 'org.junit.jupiter', 'junit-jupiter-engine').versionRef('junit')
            library('junit-params', 'org.junit.jupiter', 'junit-jupiter-params').versionRef('junit')

            version('assertj', '3.26.3')
            library('assertj', 'org.assertj', 'assertj-core').versionRef('assertj')

            bundle('junit', ['junit-api', 'junit-engine', 'junit-params', 'assertj'])

            library('jakarta-annotationApi', 'jakarta.annotation:jakarta.annotation-api:2.1.1')

            library('google-autoservice', 'com.google.auto.service:auto-service:1.1.1')
            library('joor', 'org.jooq:joor-java-8:0.9.15')

            library('lombok', 'org.projectlombok:lombok:1.18.32')

            library('snapshotTesting','io.github.origin-energy:java-snapshot-testing-junit5:4.0.8')

            library('slf4j', 'org.slf4j:slf4j-api:2.0.7')

            plugin('release', 'net.researchgate.release').version('2.8.1')
            plugin('versions', 'com.github.ben-manes.versions').version('0.51.0')
        }
    }
}