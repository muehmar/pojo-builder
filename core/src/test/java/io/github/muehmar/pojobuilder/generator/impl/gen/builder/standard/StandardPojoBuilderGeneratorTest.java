package io.github.muehmar.pojobuilder.generator.impl.gen.builder.standard;

import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;
import static io.github.muehmar.pojobuilder.snapshottesting.SnapshotUtil.writerSnapshot;
import static org.assertj.core.api.Assertions.assertThat;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import io.github.muehmar.pojobuilder.generator.Pojos;
import io.github.muehmar.pojobuilder.generator.model.Name;
import io.github.muehmar.pojobuilder.generator.model.Pojo;
import io.github.muehmar.pojobuilder.generator.model.settings.PojoSettings;
import io.github.muehmar.pojobuilder.snapshottesting.SnapshotTest;
import org.junit.jupiter.api.Test;

@SnapshotTest
class StandardPojoBuilderGeneratorTest {
  private Expect expect;

  @Test
  @SnapshotName("samplePojo")
  void standardBuilderGenerator_when_generatorUsedWithSamplePojo_then_correctOutput() {
    final Generator<Pojo, PojoSettings> generator =
        StandardBuilderGenerator.standardBuilderGenerator();
    final Writer writer =
        generator.generate(Pojos.sample(), PojoSettings.defaultSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("samplePojoAndBuilderSetMethodPrefix")
  void standardBuilderGenerator_when_samplePojoAndBuilderSetMethodPrefix_then_correctOutput() {
    final Generator<Pojo, PojoSettings> generator =
        StandardBuilderGenerator.standardBuilderGenerator();
    final Writer writer =
        generator.generate(
            Pojos.sample(),
            PojoSettings.defaultSettings().withBuilderSetMethodPrefixOpt(Name.fromString("set")),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("genericPojo")
  void standardBuilderGenerator_when_genericPojo_then_correctOutput() {
    final Generator<Pojo, PojoSettings> generator =
        StandardBuilderGenerator.standardBuilderGenerator();
    final Writer writer =
        generator.generate(Pojos.genericSample(), PojoSettings.defaultSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  void standardBuilderGenerator_when_disabledStandardBuilder_then_noOutput() {
    final Generator<Pojo, PojoSettings> generator =
        StandardBuilderGenerator.standardBuilderGenerator();
    final Writer writer =
        generator.generate(
            Pojos.sample(),
            PojoSettings.defaultSettings().withStandardBuilderEnabled(false),
            javaWriter());

    assertThat(writer.asString()).isEqualTo("");
  }
}
