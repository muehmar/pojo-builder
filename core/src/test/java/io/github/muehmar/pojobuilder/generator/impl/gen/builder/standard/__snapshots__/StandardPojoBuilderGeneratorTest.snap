genericPojo=[
java.io.IOException
java.lang.String
java.util.List
java.util.Optional

public static final class Builder0<T extends List<String>, S> {
  private final Builder<T, S> builder;

  private Builder0(Builder<T, S> builder) {
    this.builder = builder;
  }

  public Builder1<T, S> id(String id) {
    return new Builder1<>(builder.id(id));
  }
}

public static final class Builder1<T extends List<String>, S> {
  private final Builder<T, S> builder;

  private Builder1(Builder<T, S> builder) {
    this.builder = builder;
  }

  public Builder2<T, S> data(T data) {
    return new Builder2<>(builder.data(data));
  }
}

public static final class Builder2<T extends List<String>, S> {
  private final Builder<T, S> builder;

  private Builder2(Builder<T, S> builder) {
    this.builder = builder;
  }

  public OptBuilder0<T, S> andAllOptionals() {
    return new OptBuilder0<>(builder);
  }

  public Builder<T, S> andOptionals() {
    return builder;
  }

  public Customer<T, S> build() throws IOException {
    return builder.build();
  }

  /**
   * Creates a new instance of {@link CustomerBuilder} where it is guaranteed that
   * all required fields are already set.
   */
  public CustomerBuilder<T, S> toBuilder() {
    return builder.toBuilder();
  }
}

public static final class OptBuilder0<T extends List<String>, S> {
  private final Builder<T, S> builder;

  private OptBuilder0(Builder<T, S> builder) {
    this.builder = builder;
  }

  public OptBuilder1<T, S> additionalData(S additionalData) {
    return new OptBuilder1<>(builder.additionalData(additionalData));
  }

  public OptBuilder1<T, S> additionalData(Optional<S> additionalData) {
    return new OptBuilder1<>(builder.additionalData(additionalData));
  }
}

public static final class OptBuilder1<T extends List<String>, S> {
  private final Builder<T, S> builder;

  private OptBuilder1(Builder<T, S> builder) {
    this.builder = builder;
  }

  public Customer<T, S> build() throws IOException {
    return builder.build();
  }

  /**
   * Creates a new instance of {@link CustomerBuilder} where it is guaranteed that
   * all required fields are already set.
   */
  public CustomerBuilder<T, S> toBuilder() {
    return builder.toBuilder();
  }
}
]


samplePojo=[
java.io.IOException
java.lang.Integer
java.lang.String
java.util.Optional

public static final class Builder0 {
  private final Builder builder;

  private Builder0(Builder builder) {
    this.builder = builder;
  }

  public Builder1 id(Integer id) {
    return new Builder1(builder.id(id));
  }
}

public static final class Builder1 {
  private final Builder builder;

  private Builder1(Builder builder) {
    this.builder = builder;
  }

  public Builder2 username(String username) {
    return new Builder2(builder.username(username));
  }
}

public static final class Builder2 {
  private final Builder builder;

  private Builder2(Builder builder) {
    this.builder = builder;
  }

  public OptBuilder0 andAllOptionals() {
    return new OptBuilder0(builder);
  }

  public Builder andOptionals() {
    return builder;
  }

  public Customer build() throws IOException {
    return builder.build();
  }

  /**
   * Creates a new instance of {@link CustomerBuilder} where it is guaranteed that
   * all required fields are already set.
   */
  public CustomerBuilder toBuilder() {
    return builder.toBuilder();
  }
}

public static final class OptBuilder0 {
  private final Builder builder;

  private OptBuilder0(Builder builder) {
    this.builder = builder;
  }

  public OptBuilder1 nickname(String nickname) {
    return new OptBuilder1(builder.nickname(nickname));
  }

  public OptBuilder1 nickname(Optional<String> nickname) {
    return new OptBuilder1(builder.nickname(nickname));
  }
}

public static final class OptBuilder1 {
  private final Builder builder;

  private OptBuilder1(Builder builder) {
    this.builder = builder;
  }

  public Customer build() throws IOException {
    return builder.build();
  }

  /**
   * Creates a new instance of {@link CustomerBuilder} where it is guaranteed that
   * all required fields are already set.
   */
  public CustomerBuilder toBuilder() {
    return builder.toBuilder();
  }
}
]


samplePojoAndBuilderSetMethodPrefix=[
java.io.IOException
java.lang.Integer
java.lang.String
java.util.Optional

public static final class Builder0 {
  private final Builder builder;

  private Builder0(Builder builder) {
    this.builder = builder;
  }

  public Builder1 setId(Integer id) {
    return new Builder1(builder.setId(id));
  }
}

public static final class Builder1 {
  private final Builder builder;

  private Builder1(Builder builder) {
    this.builder = builder;
  }

  public Builder2 setUsername(String username) {
    return new Builder2(builder.setUsername(username));
  }
}

public static final class Builder2 {
  private final Builder builder;

  private Builder2(Builder builder) {
    this.builder = builder;
  }

  public OptBuilder0 andAllOptionals() {
    return new OptBuilder0(builder);
  }

  public Builder andOptionals() {
    return builder;
  }

  public Customer build() throws IOException {
    return builder.build();
  }

  /**
   * Creates a new instance of {@link CustomerBuilder} where it is guaranteed that
   * all required fields are already set.
   */
  public CustomerBuilder toBuilder() {
    return builder.toBuilder();
  }
}

public static final class OptBuilder0 {
  private final Builder builder;

  private OptBuilder0(Builder builder) {
    this.builder = builder;
  }

  public OptBuilder1 setNickname(String nickname) {
    return new OptBuilder1(builder.setNickname(nickname));
  }

  public OptBuilder1 setNickname(Optional<String> nickname) {
    return new OptBuilder1(builder.setNickname(nickname));
  }
}

public static final class OptBuilder1 {
  private final Builder builder;

  private OptBuilder1(Builder builder) {
    this.builder = builder;
  }

  public Customer build() throws IOException {
    return builder.build();
  }

  /**
   * Creates a new instance of {@link CustomerBuilder} where it is guaranteed that
   * all required fields are already set.
   */
  public CustomerBuilder toBuilder() {
    return builder.toBuilder();
  }
}
]